.github/workflows/ci.yml

```diff
@@ -35,43 +20,79 @@ jobs:
             os: macos-latest-xlarge
             # FIXME: some of these tests don't work very well on MacOS at the moment. Should fix
             # them at earliest convenience :)
-            flags: "--exclude integration-tests --exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
+            # Note that integration-tests are disabled at the step running them in the job below.
+            flags: "--exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
     timeout-minutes: 90
     steps:
       # Some of the tests allocate really sparse maps, so heuristic-based overcommit limits are not
       # appropriate here.
       # FIXME(#9634): remove this once the issue is resolved.
       - run: sudo sysctl vm.overcommit_memory=1 || true
       - uses: actions/checkout@v4
+
+      # Install all the required tools
       - uses: baptiste0928/cargo-install@21a18ba3bf4a184d1804e8b759930d3471b1c941
         with:
           crate: cargo-nextest
       - uses: baptiste0928/cargo-install@21a18ba3bf4a184d1804e8b759930d3471b1c941
         with:
           crate: cargo-deny
+      - uses: baptiste0928/cargo-install@21a18ba3bf4a184d1804e8b759930d3471b1c941
+        with:
+          crate: cargo-llvm-cov
+
+      # Setup the dependency rust cache and llvm-cov
       - uses: Swatinem/rust-cache@a95ba195448af2da9b00fb742d14ffaaf3c21f43
         with:
           prefix-key: "0" # change this to invalidate CI cache
           shared-key: "cargo_nextest-${{ matrix.cache_id }}"
-      - run: cargo nextest run --locked --workspace -p '*' --cargo-profile quick-release --profile ci ${{ matrix.flags }}
+      - run: cargo llvm-cov show-env | tr -d "'" >> "$GITHUB_ENV"
+
+      # Run unit tests
+      - run: cargo nextest run --locked --workspace --exclude integration-tests --cargo-profile quick-release --profile ci ${{ matrix.flags }}
         env:
           RUST_BACKTRACE: short
+      - run: cargo llvm-cov report --profile quick-release --codecov --output-path unittests.json
+      - uses: codecov/codecov-action@eaaf4bedf32dbdc6b720b63067d99c4d77d6047d
+        with:
+          files: unittests.json
+          fail_ci_if_error: true
+          flags: unittests,${{ matrix.id }}
+      # See https://github.com/taiki-e/cargo-llvm-cov/issues/292
+      - run: find target -name '*.profraw' -delete
+
+      # Run integration tests
+      - run: cargo nextest run --locked --package integration-tests --cargo-profile quick-release --profile ci ${{ matrix.flags }}
+        if: matrix.id != 'macos'
+        env:
+          RUST_BACKTRACE: short
+      - run: cargo llvm-cov report --profile quick-release --codecov --output-path integration-tests.json
+        if: matrix.id != 'macos'
+      - uses: codecov/codecov-action@eaaf4bedf32dbdc6b720b63067d99c4d77d6047d
+        if: matrix.id != 'macos'
+        with:
+          files: integration-tests.json
```

Comment:
Observation: if the coverage files were an artifact from a job (you can funnel them into a single job for upload or not – up to you) then it would be possible to download those and view them locally with whatever tooling that suits the person.

Reply:
Hmmm I think codecov format might not be the best format. Also, AFAICT the *.profraw format cannot be used without the binary too.

Considering these sentences:
> When exporting JSON, the regions, functions, branches, expansions, and summaries of the coverage data will be exported. When exporting an lcov trace file, the line-based coverage, branch coverage, and summaries will be exported.

I think JSON would be a better export, but then lcov seems to eg. be implemented in VSCode. I’m also unclear on whether it’s actually possible to convert from one format to the other.

Which format do you think would be best to upload as an artifact? (Considering the upload is fast, I think having a single upload job would probably be useless code complexity)

Reply:
I think if we’re doing this at all, we should keep the same format as what the codecov expects for the time being. If we find use-cases that would benefit from a different format we can easily add it in the future.

Reply:
Do you know of any tool able to consume codecov format, except for codecov? I don’t know of any, so I’d lean against uploading coverage as an artifact if it’s just to have it in codecov format.

Reply:
Nope, its been a long while since I used any coverage tools at all, and last time I did I IIRC just generated HTML reports with llvm’s tooling from profraws.